src/service/cid-impl.js

```diff
@@ -383,52 +390,122 @@ function setCidCookie(win, scope, cookie) {
+    return Promise.resolve(existingCookie);
   }
-
-  if (cid.externalCidCache_[scope]) {
-    return /** @type {!Promise<?string>} */ (cid.externalCidCache_[scope]);
+  if (isBackupCidExpOn && !optOutBackup) {
+    return Services.storageForDoc(ampdoc).then((storage) => {
+      const key = getStorageKey(cookieName);
+      return storage
+        .getUnexpiredValue(key, BASE_CID_MAX_AGE_MILLIS)
+        .then((backupCid) => {
+          if (backupCid) {
+            setCidCookie(win, cookieName, backupCid);
+            setCidBackup(cid.ampdoc, cookieName, backupCid);
+          }
+          return backupCid;
+        });
+    });
   }
+  return Promise.resolve();
+}
+
+/**
+ * If cookie exists it's returned immediately. Otherwise, if instructed, the
+ * new cookie is created.
+ * @param {!Cid} cid
+ * @param {!GetCidDef} getCidStruct
+ * @param {!Promise} persistenceConsent
+ * @return {!Promise<?string>}
+ */
+function getOrCreateCookie(cid, getCidStruct, persistenceConsent) {
+  const {isBackupCidExpOn, ampdoc} = cid;
+  const {win} = ampdoc;
+  const {scope, optOutBackup} = getCidStruct;
+  const cookieName = getCidStruct.cookieName || scope;
+
+  return maybeGetCidFromCookieOrBackup(cid, getCidStruct).then(
+    (existingCookie) => {
+      if (!existingCookie && !getCidStruct.createCookieIfNotPresent) {
+        return /** @type {!Promise<?string>} */ (Promise.resolve(null));
+      }
 
-  const newCookiePromise = getRandomString64(win)
-    // Create new cookie, always prefixed with "amp-", so that we can see from
-    // the value whether we created it.
-    .then((randomStr) => 'amp-' + randomStr);
-
-  // Store it as a cookie based on the persistence consent.
-  Promise.all([newCookiePromise, persistenceConsent]).then((results) => {
-    // The initial CID generation is inherently racy. First one that gets
-    // consent wins.
-    const newCookie = results[0];
-    const relookup = getCookie(win, cookieName);
-    if (!relookup) {
-      setCidCookie(win, cookieName, newCookie);
+      if (existingCookie) {
+        return /** @type {!Promise<?string>} */ (Promise.resolve(
+          existingCookie
+        ));
+      }
+
+      if (cid.externalCidCache_[scope]) {
+        return /** @type {!Promise<?string>} */ (cid.externalCidCache_[scope]);
+      }
+
+      const newCookiePromise = getRandomString64(win)
+        // Create new cookie, always prefixed with "amp-", so that we can see from
+        // the value whether we created it.
+        .then((randomStr) => 'amp-' + randomStr);
+
+      // Store it as a cookie based on the persistence consent.
+      Promise.all([newCookiePromise, persistenceConsent]).then((results) => {
+        // The initial CID generation is inherently racy. First one that gets
+        // consent wins.
+        const newCookie = results[0];
+        const relookup = getCookie(win, cookieName);
+        if (!relookup) {
+          setCidCookie(win, cookieName, newCookie);
```

Comment:
I find it a bit unintuitive to read because `getOrCreateCookie()` updates cookie/storage value. While `maybeGetCidFromCookieOrBackup()` also updates cookie/storage value.
What about updating storage value only in `getOrCreateCookie()`?

Reply:
I made it so that `maybeGetCidFromCookieOrBackup()` only retrives the CID and `getOrCreateCookie()` is the only place that sets in cookies and storage. Wdyt?

Reply:
I like this approach. Thanks